[mlir][acc] Extend PointerLikeType to provide alloc, dealloc, copy

razvanlupusoru opened the pull request:
A variable in an acc data clause operation must have a type that implements either PointerLikeType or a MappableType interface. These interfaces provide the contract that allows acc dialect and its transform passes to interact with a source dialect.

One of these requirements is ability to generate code that creates memory for a private copy and ability to initialize that copy from another variable. Thus, update the PointerLikeType API to provide the means to create allocation, deallocation, and copy. This will be used as a way to fill in privatization and firstprivatization recipes.

This new API was implemented for memref along with testing to exercise it via the implementation of a testing pass.

Commit:
[mlir][acc] Extend PointerLikeType to provide alloc, dealloc, copy

A variable in an acc data clause operation must have a type that
implements either PointerLikeType or a MappableType interface. These
interfaces provide the contract that allows acc dialect and its
transform passes to interact with a source dialect.

One of these requirements is ability to generate code that creates
memory for a private copy and ability to initialize that copy from
another variable. Thus, update the PointerLikeType API to provide
the means to create allocation, deallocation, and copy. This will
be used as a way to fill in privatization and firstprivatization
recipes.

This new API was implemented for memref along with testing to
exercise it via the implementation of a testing pass.

Commit:
Fix formatting

Commit:
Remove extraneous line

vzakhari reviewed (approved):
Looks great! Thank you, Razvan!

Commit:
Update genAllocate signature to return Value

When generating recipes, the allocated value needs to be obtainable
so that it can be threaded through to the copying and freeing
regions. Thus instead of returning true/false, return value just
for genAllocate.

clementval commented on mlir/lib/Dialect/OpenACC/IR/OpenACC.cpp:
```diff
@@ -70,6 +72,116 @@ struct MemRefPointerLikeModel
+
+    // Check if this is a static memref (all dimensions are known) - if yes
+    // then we can generate an alloca operation.
+    if (memrefTy.hasStaticShape()) {
+      return memref::AllocaOp::create(builder, loc, memrefTy).getResult();
+    }
```
no brace

Commit:
Remove extra braces

clementval reviewed (approved):
LGTM

Commit:
More brace removal to match LLVM guidelines

razvanlupusoru commented on mlir/include/mlir/Dialect/OpenACC/OpenACCTypeInterfaces.td:
```diff
@@ -70,6 +70,86 @@ def OpenACC_PointerLikeTypeInterface : TypeInterface<"PointerLikeType"> {
+    InterfaceMethod<
+      /*description=*/[{
+        Generates allocation operations for the pointer-like type. It will create
+        an allocate that produces memory space for an instance of the current type.
```
Being fixed here: https://github.com/llvm/llvm-project/pull/163861
Thank you!

razvanlupusoru commented on mlir/include/mlir/Dialect/OpenACC/OpenACCTypeInterfaces.td:
```diff
@@ -70,6 +70,86 @@ def OpenACC_PointerLikeTypeInterface : TypeInterface<"PointerLikeType"> {
+        Generates allocation operations for the pointer-like type. It will create
+        an allocate that produces memory space for an instance of the current type.
+
+        The `varName` parameter is optional and can be used to provide a name
```
Being fixed here: https://github.com/llvm/llvm-project/pull/163861
I decided to make a promise to not ignore the name when provided.
